src/Converter.cc

```diff
@@ -229,13 +240,304 @@ void Converter::ConvertImpl(tinyxml2::XMLElement *_elem,
+    {
+      Unnest(_elem);
+      // TODO(jenn) delete debug statements
+      // std::cout << "\n\n_elem after unnest:\n"
+      //             << PrintElement(_elem) << std::endl;
+    }
     else if (name != "convert")
     {
       sdferr << "Unknown convert element[" << name << "]\n";
     }
   }
 }
 
+/////////////////////////////////////////////////
+void Converter::Unnest(tinyxml2::XMLElement *_elem)
+{
+  SDF_ASSERT(_elem != NULL, "SDF element is NULL");
+
+  tinyxml2::XMLDocument *doc = _elem->GetDocument();
+
+  tinyxml2::XMLElement *nextElem = nullptr, *firstUnnestedModel = nullptr;
+  for (tinyxml2::XMLElement *elem = _elem->FirstChildElement();
+      elem;
+      elem = nextElem)
+  {
+    // break loop if reached the first unnested model
+    if (firstUnnestedModel == elem) break;
+
+    nextElem = elem->NextSiblingElement();
+    std::string elemName = elem->Name();
+
+    // skip element if not one of the following or if missing name attribute
+    if ((elemName != "frame" && elemName != "joint"
+        && elemName != "link" && elemName != "model")
+        || !elem->Attribute("name"))
+    {
+      continue;
+    }
+
+    std::string attrName = elem->Attribute("name");
+
+    size_t found = attrName.find("::");
+    if (found == std::string::npos)
+    {
+      // recursive unnest
+      if (elemName == "model")
+      {
+        // std::cout << "before recursive:\n"
+        //           << PrintElement(elem) << std::endl;
+        Unnest(elem);
+
+        // std::cout << "unnested model:\n" << PrintElement(elem) << std::endl;
+        break;
+      }
+
+      continue;
+    }
+
+    std::string newModelName = attrName.substr(0, found);
+    tinyxml2::XMLElement *newModel = doc->NewElement("model");
+    newModel->SetAttribute("name", newModelName.c_str());
+
+    if (FindNewModelElements(_elem, newModel, found + 2))
+    {
+      Unnest(newModel);
+      _elem->InsertEndChild(newModel);
+
+      // since newModel is inserted at the end, point back to the top element
+      nextElem = _elem->FirstChildElement();
+
+      if (!firstUnnestedModel)
+        firstUnnestedModel = newModel;
+      // std::cout << "newModel:\n" << PrintElement(newModel) << std::endl;
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+bool Converter::FindNewModelElements(tinyxml2::XMLElement *_elem,
+                                    tinyxml2::XMLElement *_newModel,
+                                    const size_t &_newNameIdx)
+{
+  bool unnestedNewModel = false;
+  std::string newModelName = _newModel->Attribute("name");
+
+  tinyxml2::XMLElement *elem = _elem->FirstChildElement(), *nextElem = nullptr;
+  while (elem)
+  {
+    nextElem = elem->NextSiblingElement();
+
+    std::string elemName = elem->Name();
+    std::string elemAttrName;
+
+    if (elem->Attribute("name"))
+      elemAttrName = elem->Attribute("name");
+
+    if (elemAttrName.empty() ||
+        elemAttrName.compare(0, _newNameIdx - 2, newModelName) != 0 ||
```

Comment:
nit `_newNameIdx` has semantics, but they aren't clear. Also, this term is repeated.

Perhaps state something like this?
```cc
const size_t kSeparatorLen = 2;  // "::"
size_t possibleParentNameIdx = _newNameIdx - kSeparatorLen;
```

Reply:
Alternatively, you may want to consider making `str.compare(0, posibleParentNameIdx)` a bit more robust, and use a more focused function to name splitting?

Reply:
How's this? I've updated `_newNameIdx` to be `_childNameIdx`. Also, instead of using `_newNameIdx - 2` it uses `size_t newModelNameSize = newModelName.size()`: 9ed27130e8343867e79d88dae4be3894a7f0ac82

Reply:
Excellent, thanks!